### PR TITLE
rewrite buffered line reader logic for better efficiency and testability

### PR DESCRIPTION
**Pull request by eli-darkly**
As described in https://github.com/launchdarkly/ruby-eventsource/issues/20, the current logic for parsing the response stream into lines was very inefficient in any scenario where a long line might be delivered in multiple chunks. It was also difficult to test this logic due to the way it was embedded in `SSE::Client`, and there was at least one edge case it didn't handle correctly (a CR+LF line terminator being split across two chunks).

This PR factors out the buffered line reader logic into its own class, using a more abstract model where it reads any kind of enumerator of string chunks and generates an enumerator of lines, so we can test it without using an HTTP connection. The logic is improved as follows:

* If we've scanned the current buffer and haven't yet found a line break, after reading the next chunk into the buffer we now resume scanning where we left off, instead of at the start of the buffer. This was the basic improvement suggested in https://github.com/launchdarkly/ruby-eventsource/issues/20.
* If the buffer contains multiple lines of text, previously we were calling `slice!` to cut each line out of the start of the buffer and move the remaining content up. That's potentially a lot of unnecessary string modifications. Instead, we can just keep track of where we started scanning the next line (`line_start`), emit substrings as we go along without modifying the buffer, and only remove content from the buffer after we've parsed all the lines.
* If we hit a CR terminator, set a flag so that any LF immediately following that CR will be skipped instead of treating it as a separate line terminator. Previously we were only doing this right if the CRLF appeared together in the same chunk of data; now we also do it right if the CR was at the end of a chunk.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -49,7 +49,7 @@ jobs:
       - run: gem install bundler -v 2.2.10
       - run: bundle _2.2.10_ install
       - run: mkdir ./rspec
-      - run: bundle _2.2.10_ exec rspec --format progress --format RspecJunitFormatter -o ./rspec/rspec.xml spec
+      - run: bundle _2.2.10_ exec rspec --format documentation --format RspecJunitFormatter -o ./rspec/rspec.xml spec
```
Comment (eli-darkly) [inline .circleci/config.yml:52]: This just makes the test output show the actual names of the tests rather than `....`.

File: lib/ld-eventsource/client.rb [hunk 2]
```diff
@@ -128,8 +129,6 @@ def initialize(uri,
           read: read_timeout,
           connect: connect_timeout
         })
-      @buffer = ""
-      @lock = Mutex.new
```
Comment (eli-darkly) [inline lib/ld-eventsource/client.rb:132]: The `@buffer` is now a local variable inside `BufferedLineReader.lines_from`. The `@lock` was never actually necessary, since we were only using it for synchronization within a code path that was always on the same thread.

File: lib/ld-eventsource/client.rb [hunk 3]
```diff
@@ -328,7 +287,24 @@ def read_stream(cxn)
       # it can automatically reset itself if enough time passes between failures.
       @backoff.mark_success
 
-      event_parser = Impl::EventParser.new(read_lines)
+      chunks = Enumerator.new do |gen|
```
Comment (eli-darkly) [inline lib/ld-eventsource/client.rb:290]: This is what was previously in `read_chunk_into_buffer` above. `BufferedLineReader` takes an enumerator of data chunks as a parameter, so that it can process the data abstractly without caring about whether it's from an HTTP connection or whatever. Here's the implementation of that enumerator that specifically reads from a connection.

File: lib/ld-eventsource/impl/event_parser.rb [hunk 4]
```diff
@@ -31,7 +32,6 @@ def initialize(lines)
       def items
         Enumerator.new do |gen|
           @lines.each do |line|
-            line.chomp!
```
Comment (eli-darkly) [inline lib/ld-eventsource/impl/event_parser.rb:34]: BufferedLineReader already takes care of removing the line terminator characters.

File: spec/event_parser_spec.rb [hunk 5]
```diff
@@ -5,10 +5,10 @@
 
   it "parses an event with all fields" do
     lines = [
-      "event: abc\r\n",
-      "data: def\r\n",
-      "id: 1\r\n",
-      "\r\n"
```
Comment (eli-darkly) [inline spec/event_parser_spec.rb:11]: BufferedLineReader now removes the line terminators before the lines get fed into EventParser.

File: lib/ld-eventsource/client.rb [hunk 6]
```diff
@@ -204,47 +205,14 @@ def closed?
...
-          line = read_line
-          break if line.nil?
-          gen.yield line
-        end
-      end
-    end
-    
-    def read_line
-      loop do
-        @lock.synchronize do
```
Comment (eli-darkly) [inline lib/ld-eventsource/client.rb:223]: This synchronization of access to `@buffer` using `@lock` has been removed in the refactored code - it was never necessary because the buffer is only ever accessed from one thread at a time.

File: spec/client_spec.rb [hunk 7]
```diff
@@ -36,13 +37,15 @@ def send_stream_content(res, content, keep_open:)
...
-    if !keep_open
-      wr.close
+    if keep_open
+      rd, wr = IO.pipe
+      res.body = rd
+      wr.write(content)
+      wr
+    else
+      res.body = proc { |out| out.write(content) }
+      nil
```
Comment (eli-darkly) [inline spec/client_spec.rb:47]: For unknown reasons, in JRuby sometimes using `IO.pipe` with an HTTP response and then closing the write pipe will cause the HTTP client to fail without even reading the beginning of the response. I changed it so that if we're going to close the response rather than keep it open, we don't use a pipe.

File: .circleci/config.yml (hunk 1 above)
Comment (baweaver) [reply .circleci/config.yml:52]: Definitely useful for telling which tests are failing and what's going on while debugging.

File: lib/ld-eventsource/impl/buffered_line_reader.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      # CRLF for each line; terminators are not included in the returned lines. When the
+      # input data runs out, the output enumerator ends and does not include any partially
+      # completed line.
+      #
+      # @param [Enumerator] chunks  an enumerator that will yield strings from a stream
+      # @return [Enumerator]  an enumerator that will yield one line at a time
+      #
+      def self.lines_from(chunks)
+        buffer = ""
+        position = 0
```
Comment (baweaver) [inline lib/ld-eventsource/impl/buffered_line_reader.rb:17]: Minor: `current_position` might be clearer here

File: lib/ld-eventsource/impl/buffered_line_reader.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        line_start = 0
+        last_char_was_cr = false
+
+        Enumerator.new do |gen|
+          chunks.each do |chunk|
+            buffer << chunk
+
+            loop do
+              # Search for a line break in any part of the buffer that we haven't yet seen.
```
Comment (baweaver) [inline lib/ld-eventsource/impl/buffered_line_reader.rb:26]: Like the comments here, was a bit hard to follow without them before.

File: spec/buffered_line_reader_spec.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  def make_tests(name, input_line_chunks, expected_lines)
+    [{
+      name: "#{name}: one chunk per line",
+      chunks: input_line_chunks,
+      expected: expected_lines
+    }].concat(
+      (1..4).map do |size|
+        ({
+          name: "#{name}: #{size}-character chunks",
+          chunks: input_line_chunks.join().chars.to_a.each_slice(size).map { |a| a.join },
```
Comment (baweaver) [inline spec/buffered_line_reader_spec.rb:13]: Fun thing with Ruby, `join` will go down collections:

```ruby
['a', 'b', ['c', ['d']]].join
# => "abcd"
```

So you could potentially do something like this:

```ruby
input_line_chunks.flat_map(&:chars).each_slice(size).join
```

Comment (eli-darkly) [reply spec/buffered_line_reader_spec.rb:13]: But that's not what I want. The result of this expression is not supposed to be a single string, but an array of strings.

The intention here is to test for edge cases that can result from lines or line terminators being split differently across chunks as the chunks are read into the buffer. If the original tests specifies `input_line_chunks` as `-`["line1\r\n", "line2\r\n"]`, we will test with exactly those chunks, but we'll also test with them split up differently: for instance when `size` is 2, we want the chunks to be delivered as `["li", "ne", "1\r", "\nl", "in", "e2", "\r\"]`.

File: lib/ld-eventsource/impl/buffered_line_reader.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
+
+module SSE
+  module Impl
+    class BufferedLineReader
```
Comment (baweaver) [inline lib/ld-eventsource/impl/buffered_line_reader.rb:4]: For this particular case and at this point in time this likely won't be an issue, but singleton-classes can be a bit harder to work with and end up with run-on logic as things get added to them.

File: lib/ld-eventsource/impl/buffered_line_reader.rb [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      #
+      def self.lines_from(chunks)
+        buffer = ""
+        position = 0
+        line_start = 0
+        last_char_was_cr = false
+
+        Enumerator.new do |gen|
+          chunks.each do |chunk|
+            buffer << chunk
```
Comment (baweaver) [inline lib/ld-eventsource/impl/buffered_line_reader.rb:23]: Buffer length may get slightly costly as the buffer grows. I don't think this is a concern up until you get above 200Mb or so, but might be worth considering a running numerical tracker if that bottlenecks in the future.

File: spec/event_parser_spec.rb [hunk 13]
```diff
@@ -70,12 +70,12 @@
...
-      "\r\n",
-      "data: ghi\r\n",
-      "\r\n"
+      "event: abc",
+      "data: def",
+      "id: 1",
+      "",
+      "data: ghi",
+      ""
     ]
```
Comment (baweaver) [inline spec/event_parser_spec.rb:79]: One thing that might come in handy for these is squiggly heredocs which deindent to the least indented field:

```ruby
lines = <<~EVENT.lines.map(&:chomp)
  event: abc
  data: def
  id: 1
  
  data: ghi

EVENT
# => ["event: abc", "data: def", "id: 1", "", "data: ghi", ""]
```

File: lib/ld-eventsource/client.rb (hunk 3 above)
Comment (baweaver) [reply lib/ld-eventsource/client.rb:302]: Interestingly `HTTP.rb` has a response body (`HTTP::Response::Body#each`) which you can treat as an `Enumerable` entity:

https://github.com/httprb/http/blob/main/lib/http/response/body.rb#L36

```ruby
# Iterate over the body, allowing it to be enumerable
def each
  while (chunk = readpartial)
    yield chunk
  end
end
```

...which may reduce the amount of code necessary to work with  chunks down to something like:

```ruby
chunks = Enumerator.new do |gen|
  begin
    cxn.each do |chunk|
      break if @stopped.value
      break if chunk.nil?

      gen.yield chunk
    end
  rescue HTTP::TimeoutError 
    # For historical reasons, we rethrow this as our own type
    raise Errors::ReadTimeoutError.new(@read_timeout)
  end
end
```

...though I really wish they'd approve and merge the block-scoped rescues so you could do this instead:

```ruby
chunks = Enumerator.new do |gen|
  cxn.each do |chunk|
    break if @stopped.value
    break if chunk.nil?

    gen.yield chunk
  rescue HTTP::TimeoutError 
    # For historical reasons, we rethrow this as our own type
    raise Errors::ReadTimeoutError.new(@read_timeout)
  end
end
```

File: spec/buffered_line_reader_spec.rb [hunk 14]
```diff
@@ -0,0 +1,62 @@
+require "ld-eventsource/impl/buffered_line_reader"
+
+def tests_for_terminator(term, desc)
+  def make_tests(name, input_line_chunks, expected_lines)
+    [{
+      name: "#{name}: one chunk per line",
+      chunks: input_line_chunks,
+      expected: expected_lines
+    }].concat(
+      (1..4).map do |size|
```
Comment (baweaver) [inline spec/buffered_line_reader_spec.rb:10]: minor: You could do `4.times.map` there

File: spec/buffered_line_reader_spec.rb [hunk 15]
```diff
@@ -0,0 +1,62 @@
+require "ld-eventsource/impl/buffered_line_reader"
+
+def tests_for_terminator(term, desc)
+  def make_tests(name, input_line_chunks, expected_lines)
```
[the remark this answers is not in the document]
Comment (baweaver) [reply spec/buffered_line_reader_spec.rb:4]: Also super petty, but why `3rd line` and `third line`?

File: lib/ld-eventsource/impl/buffered_line_reader.rb (hunk 11 above)
Comment (eli-darkly) [reply lib/ld-eventsource/impl/buffered_line_reader.rb:4]: There's a specific reason that this code was moved to its own class: to allow it to be unit-tested outside of the context of SSE::Client. I'd rather not get into hypothetical concerns of every possible way someone might overcomplicate this code in the future.

File: lib/ld-eventsource/impl/buffered_line_reader.rb (hunk 12 above)
Comment (eli-darkly) [reply lib/ld-eventsource/impl/buffered_line_reader.rb:23]: I'm not sure what you're getting at. The buffer is only going to grow large if either 1. a large amount of data containing multiple lines arrives from the socket before we've had a chance to parse it— in which case we will throw it away as soon as we have parsed those lines, which we would have to do anyway— or 2. there is one very long line of event data, in which case we are going to need to store that data as a single string in the SSE event anyway; there would be no savings in breaking it up if we're simply going to re-concatenate it later.

File: spec/event_parser_spec.rb (hunk 13 above)
Comment (eli-darkly) [reply spec/event_parser_spec.rb:79]: The intention of this PR isn't to rewrite the entire unit test codebase, but if I were going to do so, I still wouldn't want to do this. In a test case where the exact content of each string in the list is important, I prefer to represent that unambiguously with basic syntax, rather than require the reader to play computer in order to imagine what the actual data will end up being after the transformation.

File: spec/buffered_line_reader_spec.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (eli-darkly) [reply spec/buffered_line_reader_spec.rb:4]: It's useful in catching edge cases to have input data of different lengths, which "first line" and "third line" would not be.

File: spec/buffered_line_reader_spec.rb [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+      ["first line" + term + "second line" + term + "third",
+       " line" + term + "fourth line" + term],
+      ["first line", "second line", "third line", "fourth line"])
+  ].flatten
+end
+
+describe SSE::Impl::BufferedLineReader do
+  subject { SSE::Impl::BufferedLineReader }
+
+  terminators = {
```
Comment (louis-launchdarkly) [inline spec/buffered_line_reader_spec.rb:46]: Is it possible for SSE to have multiple kinds of terminators in a message? Even so, the tests that are here maybe enough to verify that case work work already even if that is possible.

File: spec/buffered_line_reader_spec.rb (hunk 10 above)
Comment (eli-darkly) [reply spec/buffered_line_reader_spec.rb:13]: @louis-launchdarkly OK - the reason I didn't leave a comment is that the name of the test is already "2-character chunks", etc. which I thought described the intention of making 2-character chunks, but I can add one.

File: spec/buffered_line_reader_spec.rb (hunk 16 above)
Comment (eli-darkly) [reply spec/buffered_line_reader_spec.rb:46]: It's unlikely that any server is implemented that way, but the spec does allow it, so it's reasonable to add some tests for that.
